apps/framework-cli/src/infrastructure/redis/redis_client.rs

```diff
@@ -0,0 +1,541 @@
+        let incomplete_queue = format!("{}::{}::mqincomplete", REDIS_KEY_PREFIX, self.service_name);
+
+        if success {
+            let _: () = self
+                .connection
+                .lock()
+                .await
+                .lrem(&in_progress_queue, 0, message)
+                .await
+                .context("Failed to mark queue message as successful")?;
+        } else {
+            let mut pipeline = redis::pipe();
+            pipeline
+                .lrem(&in_progress_queue, 0, message)
+                .rpush(&incomplete_queue, message);
+
+            let _: () = pipeline
+                .query_async(&mut *self.connection.lock().await)
+                .await
+                .context("Failed to mark queue message as incomplete")?;
+        }
+        Ok(())
+    }
+
+    pub fn start_periodic_tasks(&mut self) {
+        info!("Starting periodic tasks");
+
+        self.presence_task = Some(tokio::spawn({
+            let mut presence_client = self.clone();
+            async move {
+                let mut interval = interval(Duration::from_secs(PRESENCE_UPDATE_INTERVAL));
+                loop {
+                    interval.tick().await;
+                    if let Err(e) = presence_client.presence_update().await {
+                        error!("Error updating presence: {}", e);
+                    }
+                }
+            }
+        }));
+
+        self.lock_renewal_task = Some(tokio::spawn({
+            let mut lock_renewal_client = self.clone();
+            async move {
+                let mut interval = interval(Duration::from_secs(LOCK_RENEWAL_INTERVAL));
+
+                // First attempt to gain leadership
+                match lock_renewal_client.attempt_leadership().await {
+                    Ok(is_leader) => {
+                        if is_leader {
+                            info!("Successfully gained leadership");
+                        } else {
+                            info!("Failed to gain leadership initially");
+                        }
+                    }
+                    Err(e) => error!("Error attempting leadership: {}", e),
+                }
+
+                loop {
+                    interval.tick().await;
+                    if lock_renewal_client.is_current_leader() {
+                        if let Err(e) = lock_renewal_client.renew_lock().await {
+                            error!("Error renewing lock: {}", e);
+                        }
+                    } else {
+                        // Attempt to gain leadership if not currently the leader
+                        if let Err(e) = lock_renewal_client.attempt_leadership().await {
+                            error!("Error attempting leadership: {}", e);
+                        }
+                    }
+                }
+            }
+        }));
+
+        info!("Periodic tasks started");
+    }
+
+    pub async fn stop_periodic_tasks(&mut self) -> Result<()> {
```

Comment:
This doesn't have to be async too right?

Reply:
That's right.